### DISCUSSION TITLE
Elipse fitting of grain from data sheet

### DISCUSSION THREAD

**Discussion by slzrd99** [mtex-toolbox/mtex#1093] [Ask Anything]
Hello,
I have a set data with the euler angles, element ID numbers ( X Y Z) , Grain ID, and Phase ID. Is there any command to fit an elipse for each grain with its equivalent grain size and obtain the main parameters such as long axis, short axis and omega ( angle giving trend to the long axis)?. Somethig similar to the fitElipse command but that works instead of with directly EBSD data with the said input data file.

Thank you very much.

**Comment by kilir** [comment]
Hi,
can you elaborate on the data, especially what " element ID numbers ( X Y Z)" looks like?
Cheers,
Rüdiger

**Reply by slzrd99** [reply]
Yes so X, Y, Z are integer indices into the Image geometry. So this numbers will define which part of an RVE is part of the grain. So if my structure has a mesh size of 64 the 3D volume of the element will be 64x64x64. So the element ID number of each grain will be something like (1,1,1), (2,1,1); for the first layer until (1,1,64) and then it would sart with (1,2,1),(2,2,1) etc. and like this until the last one which will be (64,64,64). So if we gather the grains with the same Grain ID number ( which is the feature_ID) and taking into account the element ID numbers, we should be able to see the overall shape of the grain, however I am not able to fit an ellipse with the appropriate grain size, which is calculated using the Grain ID.

**Reply by kilir** [reply]
Ok. If you can derive lists of vertices (x,y) for each grain, you can simply slightly modify the Mtex function `principalComponents()` for this task.
Cheers,
Rüdiger

**Reply by slzrd99** [reply]
but it seems that in order to input data to the principalComponents() comand, first the grain2d comman must be used which need the input of EBSD data that I do not have in that format, so I am stuck at that point as I feel like I lack information in order to use that command

**Reply by kilir** [reply]
No need fro grains, that's what I meant with modifying `principalComponents()`, if you look at the function, `V(poly{k})` simply returns the list of grain vertices. Can you derive lists of vertices from your data?

**Reply by slzrd99** [reply]
Yes I have derived the lists of vertices but I still do not understand how to do the fitting.

**Reply by kilir** [reply]
This is taken from `principalComponents()` amongst others and works on a nx2 list of xy coordinates (sorted vertices), this should work for your lists.

```
%compute centroid
dF = (xy(1:end-1,1).*xy(2:end,2)-xy(2:end,1).*xy(1:end-1,2));
cx = (xy(1:end-1,1) +xy(2:end,1)).*dF;
cy = (xy(1:end-1,2) +xy(2:end,2)).*dF;

ndx = 1:length(xy)-1;
A = sum(dF(ndx));
c = [sum(cx(ndx)) / 3 / A  sum(cy(ndx)) / 3 / A];
    
% center polygon in centroid
xy = xy - c; 

% compute length of line segments
dist = sqrt(sum((xy(1:end-1,:) - xy(2:end,:)).^2,2));
dist = 0.5*(dist(1:end) + [dist(end);dist(1:end-1)]);
  
% weight vertices according to half the length of the adjacent faces
xy = xy(1:end-1,:) .* [dist,dist] .* sum(xy(1:end-1,:).^2,2).^(0.25);
      
  % compute eigen values and vectors
[ew, omega] = eig2(xy' * xy);
    
  % halfaxes are square roots of the eigenvalues
b = sqrt(ew(1)); a = sqrt(ew(2));

% scale half axes
scaling = sqrt(A/2 ./ a ./ b ./pi);
a = a .* scaling
b = b .* scaling
```
